### ISSUE TITLE
Create a CodeQL template in arcade eng/common

### ISSUE THREAD

**Issue by dkurepa** [dotnet/arcade#13495]
Currently we're using the same yaml for running the code-ql weekly pipeline in 3 of our repos, arcade-services, dnceng, and dnceng-shared. In order to be able to makes changes/fixes in only one place, it might be worth it to create a template for this in arcade eng/common as @premun suggested in this PR https://github.com/dotnet/dnceng-shared/pull/8#discussion_r1192046128

**Comment by premun** [comment]
I think the list is longer - xharness, xliff-tasks, arcade etc.

**Comment by dkurepa** [comment]
PR: https://github.com/dotnet/arcade/pull/13526

**Comment by dkurepa** [comment]
In the PR comments, we decided that arcade is not the correct place for this template. We want to explore the idea of having these **dnceng only** templates in a different repo, like `dotnet/dnceng-shared`, and using that repo as a resource in the pipelines: https://learn.microsoft.com/en-us/azure/devops/pipelines/process/resources?view=azure-devops&tabs=schema

**Comment by riarenas** [comment]
> @riarenas what do you think would be the associated extra cost? Maybe we could sort of make it self-service as some other components in the Arcade?

I'm not really sure. I know that we made the explicit decision to not offer a template. @garath will probably have some more details on this.

Another thing to keep in mind is that dependency flow and the eng/common folder sync were built in the very early days of yaml templates in Azure Pipelines. I think resources weren't really a thing back then so there might be other things that make more sense to use. If we can figure something out for the dnceng repos, we might have a pattern for improvements to the eng/common situation. (there have been a few issues in the past for figuring out improvements or more drastic changes to the current sync method)

**Comment by tkapin** [comment]
> I'm not really sure. I know that we made the explicit decision to not offer a template. @garath will probably have some more details on this.

@garath - would you know more about this info from @riarenas? Since CodeQL offers static code analysis, I assume many repos could benefit from this. Whoy wouldn't we provide a template for this? Are there any security concerns here?

**Comment by garath** [comment]
I opted to not write a one-codeql-template-fits-all because I couldn't really find a way to make it work reasonably. The CodeQL 3000 plugin takes all build data between the Init and Finalize tasks. If the build is sufficiently complicated this may be too large of a scope so a custom path would be required. If the build is simple, the CodeQL 3000 plugin is really only two, plain AzDO tasks, and even those may be optional if building from the default branch.

**Comment by garath** [comment]
For DNCEng repositories, were I to set it all up again, I would try to simplify even more by considering not adding a separate YAML file and instead just adding the tasks in the primary build YAML.

**Comment by premun** [comment]
> If the build is sufficiently complicated this may be too large of a scope so a custom path would be required.

Don't we do the same for things like `jobs.yml` where if the repo is sufficiently complicated goes their own way but has the possibility of using tools from Arcade?
We don't necessarily need everyone to use the template but the possibility would be there?

> the CodeQL 3000 plugin is really only two, plain AzDO tasks

Seems like it's more than that as it's also some variables, installing .NET for the tool and so on?
https://github.com/dkurepa/dnceng-shared/blob/dkurepa/YamlTemplate/eng/templates/run-codeql.yml

**Comment by dkurepa** [comment]
> Seems like it's more than that as it's also some variables, installing .NET for the tool and so on?

I think it's the variables and the two AzDO tasks, helix-service is a good representation of this https://dev.azure.com/dnceng/internal/_git/dotnet-helix-service?path=/azure-pipeline-codeql.yaml



> For DNCEng repositories, were I to set it all up again, I would try to simplify even more by considering not adding a separate YAML file and instead just adding the tasks in the primary build YAML.

I imagine with this approach we'd still only want to run this on a weekly basis? We could add the scheduled triggers to the main pipelines and have something like `${{ if eq(variables['Build.Reason'], 'Schedule') }}: task: CodeQL3000Init@0`.
I think the downside of this approach is that we'd be losing some visibility that we have with a separate pipeline

**Comment by dkurepa** [comment]
The GitHub issue verification also something we use in all of our repos. If we decide to use dnceng-shared for our templates, we could put it there too

**Comment by garath** [comment]
> I think the downside of this approach is that we'd be losing some visibility that we have with a separate pipeline

I only suggest using the same YAML. There would still be two pipelines. Though, the CodeQL 3000 plugin does have some specific features that try to eliminate the need for multiple pipelines, and those might be good enough.

**Comment by garath** [comment]
> I imagine with this approach we'd still only want to run this on a weekly basis? We could add the scheduled triggers to the main pipelines and have something like ${{ if eq(variables['Build.Reason'], 'Schedule') }}: task: CodeQL3000Init@0.

Just a tiny thing so more of my thinking is here, I considered it important to be able to run the CodeQL pipeline manually (so one could verify changes) so a simple "if scheduled" conditional wouldn't quite be enough.

**Comment by garath** [comment]
> Seems like it's more than that as it's also some variables, installing .NET for the tool and so on?

Not quite. The additional Tasks are for build itself. I would not include them in a generic template.

The pipeline variables are just exposing a bit of control I thought useful when resolving CodeQL actions.

The CodeQL 3000 plugin has no other requirements. And indeed, as I mentioned, if you only care about running CodeQL against the default branch, you don't even need to explicitly use the Init and Finalize tasks. They are auto-injected into pipelines by policy (just like component governance).

To be clear, I am open to adding a template.

**Comment by premun** [comment]
@garath apologies, I see your points now, I didn't know the steps in the template are not required for CodeQL.
We should take those out if this was to be in Arcade, yeah.

If things are like this, I am starting to lean towards this not being in Arcade but in dnceng.
